### PR TITLE
add additional services

### PR DESCRIPTION
**Pull request by phantomnat**
- getRemaining -> return up to 10 available seats
- reserve -> reserve the seat for up to X seconds
- confirm -> confirm booking
- cancel -> cancel the rsvp
add unit test
update apidoc

### CODE REVIEW

**Comment by varokas** [comment]
thanks! this looks great. I'll play with it tonight my time and merge.

File: lib/book.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        // seat valid and available
+    } else if (
+        seatNo in seats && 
+        seats[seatNo].status === bookingStatus.reserved &&
+        now >= seats[seatNo].reservedExpiredTime
+    ) {
+        // reserve expire
+    } else {
+        // seatNo not valid or not available
+        // let get an available seat for user
```
Comment (varokas) [inline lib/book.js]: For the new rule, we no longer need this logic to grab a seat. Let's make "seat" a required parameter for the "/book" API.

We will let the client refresh /bookings repeatedly to contend for resource. Just like real world =)

File: lib/book.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                    seat.status === bookingStatus.reserved &&
+                    now >= seat.reservedExpiredTime
+                )
+            ) {
+                seatNo = seatNos[i]
+                break
+            }
+        }
+    }
+    if (!(seatNo in seats)) {
```
Comment (varokas) [inline lib/book.js]: again, no need for this. If the seatNo happens to reserved. we will simply returns false

File: index.js [hunk 3]
```diff
@@ -1,23 +1,44 @@
...
+ *        "A1",
+ *        "A2",
+ *        ...
+ *      ]
+ *    }
+ */
+app.get('/remaining/:noOfSeats*?', function(req, res) {
+    const result = book.getRemaining(req.params.noOfSeats)
+    res.json({ seats: result})
+})
```
Comment (varokas) [inline index.js]: to make this even simpler, we could just hard-coded number 10 as in rule. no need to have a parameter here.

File: index.js [hunk 4]
```diff
@@ -26,35 +47,85 @@ bookings.push({"round": currentRound, "seats": []})
...
-    res.send({"round": currentRound, "seat": seat})
-
-    bookings[bookings.length - 1].seats.push(seat)
-
-    currentCol++
-    if (currentCol > lastCol) {
-        currentCol = 1
-        currentRow = String.fromCharCode(currentRow.charCodeAt() + 1);
+app.post('/book', function(req, res) {
+    const result = book.reserve(req.body.seat)
```
Comment (varokas) [inline index.js:69]: See comment below about `reserve(seat)`. We should just reject request with `seat == null`

**Comment by varokas** [comment]
@phantomnat - i tried it and this works really good. Thanks for your contribution. Let me know what you think about my comments. 

@mahasak please review (or add tossapon to help review)

File: lib/book.js (hunk 2 above)
Comment (tnuanchuay) [inline lib/book.js]: Should we go find the seat next round if there is no seat available in this round?
It seems like he/she need to confirm the last seat and then we can go further to next round.

We should have rules for this, if last seat isn't confirmed yet and customers are coming. they can/can't get next round seat. @varokas

**Review by tnuanchuay — APPROVED** [review]
Looks good to me, but have concern

**Comment by phantomnat** [comment]
@tspn I think it would make sense if we're selling thing like 10K+ tickets for UEFA Champions League Final which is no advancing round.

I'll update the pr after we finish the discussion about this.

**Comment by varokas** [comment]
@phantomnat  - sorry i forgot to submit review

File: lib/book.js (hunk 2 above)
Comment (varokas) [reply lib/book.js]: which is fine. we want to have some contention going on. but in light of this (https://github.com/bigbearsio/strength/issues/6#issuecomment-391597105) we probably will just remove rounds altogether. 

Let's skip this concern and merge this after other comments are addressed. Then we can do another PR to remove rounds?

**Comment by phantomnat** [comment]
lets check this again @varokas

File: index.js (hunk 4 above)
Comment (phantomnat) [reply index.js:69]: `reserve()` can handle `null` and will return `false`

**Comment by varokas** [comment]
looks good. tested and works perfectly. Thanks a lot for contributing

**Comment by varokas** [comment]
@mahasak @tspn  - im merging this

**Pull request merged**